File: CarePackage/Software/NvidiaGeforceExperience.cs
namespace CarePackage.Software;

public class NvidiaGeforceExperience : BaseSoftware
{
    public override string           Key            { get; set; } = "nvidia-geforce-experience";
    public override string           Name           { get; set; } = "NVIDIA GeForce Experience";
    public override SoftwareCategory Category       { get; set; } = SoftwareCategory.Peripheral;
    public override string           DownloadName   { get; set; } = "GeForce_Experience.exe";
    public override bool             IsArchive      { get; set; } = false;
    public override bool             ShouldCacheUrl { get; set; } = true;
    public override bool             RequiresAdmin  { get; set; } = true;
    public override Bitmap           Icon           { get; set; } = Resources.Icons.nvidia_geforce_experience;
    public override string           Homepage       { get; set; } = "https://nvidia.com/en-us/geforce/geforce-experience";

    private readonly HttpClient _http;

    public NvidiaGeforceExperience(IServiceProvider services)
    {
        _http = services.GetKeyedService<HttpClient>("MimicBrowser")!;
    }
    
    public override async Task<string> GetDownloadUrlAsync(CancellationToken ct)
    {
        var res = await _http.GetAsync("https://www.nvidia.com/en-us/geforce/geforce-experience/", ct);
        
        res.EnsureSuccessStatusCode();

        var downloadUrlPattern = new Regex(@"https://us\.download\.nvidia\.com/GFE/GFEClient/\d{1,}\.\d{1,}\.\d{1,}\.\d{1,}/GeForce_Experience_v\d{1,}\.\d{1,}\.\d{1,}\.\d{1,}\.exe");
        var html               = await res.Content.ReadAsStringAsync(ct);
        var match              = downloadUrlPattern.Match(html);
        
        DownloadUrlResolveException.ThrowUnless(match.Success);
        
        return match.Groups[0].Value;
    }
}

File: src/lib/software/nvidia_geforce_experience.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class NvidiaGeForceExperience(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'nvidia-geforce-experience'
        self.name = 'NVIDIA GeForce Experience (Deprecated)'
        self.category = SoftwareCategory.Peripheral
        self.download_name = 'GeForce_Experience.exe'
        self.is_archive = False
        self.should_cache_url = True
        self.requires_admin = True
        self.icon = 'nvidia.png'
        self.homepage = 'https://nvidia.com/en-us/geforce/geforce-experience'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        download_url_pattern = compile(r'https://us\.download\.nvidia\.com/GFE/GFEClient/\d+\.\d+\.\d+\.\d+/GeForce_Experience_v\d+\.\d+\.\d+\.\d+\.exe')
        html = reply.readAll().data().decode()
        match = download_url_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(match.group(0))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest('https://www.nvidia.com/en-ph/geforce/geforce-experience/'))
